### DISCUSSION TITLE
Setting the "Scope" property to "perUserOrMachine" results in an unexpected INSTALLFOLDER path

### DISCUSSION THREAD

**Discussion by Arkarin0** [wixtoolset/issues#7624] [Questions] [answered]
If this issue is a bug:

* Which version of WiX are you building with?

> WixToolset.UI.wixext 4.0.1

* Which version of Visual Studio are you building with (if any)?

> Microsoft Visual Studio Community 2022 (64-Bit) - Current
Version 17.4.4

* Which version of the WiX Toolset Visual Studio Extension are you building with (if any)?

> HeatWave for VS2022 V1.0.1

* Which version of .NET are you building with?

> 6.0.102 x64/x86

* If the problem occurs when installing your packages built with WiX, what is the version of Windows the package is running on?

> Windows 10 Pro - 10.0.19045 Build 19045
> Windows ® Installer. V 5.0.19041.2193 

* Describe the problem and the steps to reproduce it.

> Setting the "Scope" property in the Package node to "perUserOrMachine" results in an unexpected INSTALLFOLDER path.
>
>Steps to reproduction:
>1. Create a new msi package project using VS and heatwave.
>2. Add the following nuget package to the Project: WixToolset.UI.wixext V4.0.1
>3. add the `ui="http://wixtoolset.org/schemas/v4/wxs/ui` namespace to the package.wxs file.
>4. Then add the following line felow the feature element (used for easier debuging). `<ui:WixUI Id="WixUI_InstallDir" InstallDirectory="INSTALLFOLDER"/>`
>5. Set the Scope property in the Package node. `Scope="perUserOrMachine"`
>6. build the package
>7. run the installer. press yes till the "Choose Install location" Dialog appears.

* Describe the behavior you expected and how it differed from the actual behavior.

> Setting the Scope to "perUser" or "perMachine" sets the variable INSTALLFOLDER to "C:\Program Files\[Manufacturer] [ProductName]\". </br>Setting the Scope to "perUserOrMachine" sets the variable INSTALLFOLDER to "C:\Users\[CurrentUser]\AppData\Local\Programs\\[Manufacturer] [ProductName]\".</br></br>I would expect the Path to match the Path of "perUser" or "perMachine" but not to neither match one of them. "perUserOrMachine" implies that it is either a "perUser" or "perMachine" installation so the Path should result to the corrosponding paths for "perUSer" and "perMachine".

thanks for your great work. :-)

**Comment by barnson** [comment, marked as answer]
FWIW, there is no WixUI dialog set that supports `perUserOrMachine`.

**Reply by ggeurts** [reply]
I just was bitten by the lack of WixUI support for the perUserOrMachine setting as well. The WixUI_Advanced dialog does not set the install location to program files if a per machine installation is chosen in the UI.

**Comment by Arkarin0** [comment]
So its only an issue when used in combination with the UI?

I am correct in the assumption that

- the `scope` property of the `package` should only be set in headless applications.
- for UI application the [InstallScopeDlg.wxs](https://github.com/wixtoolset/wix/blob/develop/src/ext/UI/wixlib/InstallScopeDlg.wxs) dialog should be used.
- Is there an example of a proper implementation for both cases?

**Reply by barnson** [reply]
I'm not sure what's required. InstallScopeDlg was created before `perUserOrMachine` support.

**Reply by Arkarin0** [reply]
if we can agree on  the desired behavior. I am interested to try to fix it.

**Comment by Arkarin0** [comment]
after a little bit of research i also found, that setting `scope` to `perUser` should result in a displayed path like
`C:\Users\\[CurrentUser]\AppData\Local\Programs\\[Manufacturer] [ProductName]`

So to summarize, I would expect:
- setting `scope` to `perUser` should result [InstallDirDlg.wxs](https://github.com/wixtoolset/wix/blob/develop/src/ext/UI/wixlib/InstallDirDlg.wxs) displaying a path like `C:\Users\[CurrentUser]\AppData\Local\Programs\[Manufacturer] [ProductName]`
- setting `scope` to `perMachine` should result [InstallDirDlg.wxs](https://github.com/wixtoolset/wix/blob/develop/src/ext/UI/wixlib/InstallDirDlg.wxs) displaying a path like `C:\Program Files\[Manufacturer] [ProductName]`
- setting `scope` to `perUserOrMachine` and selecting a `perUser` installation should result in the `perUser` behavior.
- setting `scope` to `perUserOrMachine` and selecting a `perMachine` installation should result in the `perMachine` behavior.
- consistent behavior between headed and headless installations. (i have not tested yet if the behavior is correct when using headless installations)

**Reply by barnson** [reply]
That sounds right. (ProgramFiles6432Folder should take care of the right Program Files directory.)

**Comment by Arkarin0** [comment]
I created a fork with an issue [branch](https://github.com/Arkarin0/wix/tree/e0641f1479c6f67f6ec119d448b66a74ad5d6b88).

I will try me best to investigate and fix the issue.
